### PR TITLE
[SLPVectorizer][X86] Free load cost for stores with constant pointers

### LINKED ISSUE #111126: Regression from clang 16: missed vectorization in simple array initialization
reduced code:
https://godbolt.org/z/3MYzE1v7M
```cpp
long arr [20];

void f() {
    for (int i = 0; i < 20; i += 1) 
    {
        arr [i] = 1;
    }
}
```

clang -O3:
```asm
f():
        mov     qword ptr [rip + arr], 1
        mov     qword ptr [rip + arr+8], 1
        mov     qword ptr [rip + arr+16], 1
        mov     qword ptr [rip + arr+24], 1
        mov     qword ptr [rip + arr+32], 1
        mov     qword ptr [rip + arr+40], 1
        mov     qword ptr [rip + arr+48], 1
        mov     qword ptr [rip + arr+56], 1
        mov     qword ptr [rip + arr+64], 1
        mov     qword ptr [rip + arr+72], 1
        mov     qword ptr [rip + arr+80], 1
        mov     qword ptr [rip + arr+88], 1
        mov     qword ptr [rip + arr+96], 1
        mov     qword ptr [rip + arr+104], 1
        mov     qword ptr [rip + arr+112], 1
        mov     qword ptr [rip + arr+120], 1
        mov     qword ptr [rip + arr+128], 1
        mov     qword ptr [rip + arr+136], 1
        mov     qword ptr [rip + arr+144], 1
        mov     qword ptr [rip + arr+152], 1
        ret
```

expected code (clang 15 or gcc):
```asm
f():
        movaps  xmm0, xmmword ptr [rip + .LCPI0_0]
        movaps  xmmword ptr [rip + arr], xmm0
        movaps  xmmword ptr [rip + arr+16], xmm0
        movaps  xmmword ptr [rip + arr+32], xmm0
        movaps  xmmword ptr [rip + arr+48], xmm0
        movaps  xmmword ptr [rip + arr+64], xmm0
        movaps  xmmword ptr [rip + arr+80], xmm0
        movaps  xmmword ptr [rip + arr+96], xmm0
        movaps  xmmword ptr [rip + arr+112], xmm0
        movaps  xmmword ptr [rip + arr+128], xmm0
        movaps  xmmword ptr [rip + arr+144], xmm0
        ret

```

### PR DESCRIPTION
**Pull request by antoniofrighetto**
When estimating the cost for stores of constant buildvectors, do not take into account the cost of the additional load to materialize a vector from a constant pool when dealing with a constant pointer. In such cases, the load is avoided in the first place, as the only operations required simply involve computing the address of the constant (`ip+base_addr+offset`) and the store itself.

Fixes: https://github.com/llvm/llvm-project/issues/111126.

### CODE REVIEW

**Comment by antoniofrighetto** [comment]
FWIW, the load emitted by clang-15 (https://godbolt.org/z/nbofoe6zM) shouldn't be related with this, and in any cases, it just seems suboptimal codegen (GCC doesn't emit any load wrt to this instance).

**Comment by alexey-bataev** [comment]
> @alexey-bataev, now that I look better at it, I think that's the load you were referring to in [0e7ed32](https://github.com/llvm/llvm-project/commit/0e7ed32c71362f3547329c6ee8573a8bc191f58a)? Maybe we should remove it all?

Sorry, do not quite understand. What's better to remove?

File: llvm/lib/Target/X86/X86TargetTransformInfo.cpp [hunk 1]
```diff
@@ -5157,8 +5157,9 @@ InstructionCost X86TTIImpl::getMemoryOpCost(unsigned Opcode, Type *Src,
 
   InstructionCost Cost = 0;
 
-  // Add a cost for constant load to vector.
-  if (Opcode == Instruction::Store && OpInfo.isConstant())
+  // Add a cost for constant load to vector, if pointer is not a constant.
+  if (auto *SI = dyn_cast_or_null<StoreInst>(I);
+      SI && !isa<Constant>(SI->getPointerOperand()) && OpInfo.isConstant())
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86TargetTransformInfo.cpp:5162]: We're trying to account for the cost of the constant vector load, but you want to skip it if the store address is constant?

Comment (RKSimon) [reply llvm/lib/Target/X86/X86TargetTransformInfo.cpp:5162]: Looking at #111126 - shouldn't we be checking to see if `OpInfo.isUniform()` and adjusting the Type accordingly?

It looks like the test case thinks it has to load a `<20 x i64> splat(i64 1)` constant when it just has to load `<2 x i64> splat(i64 1)`

Comment (antoniofrighetto) [reply llvm/lib/Target/X86/X86TargetTransformInfo.cpp:5162]: Sorry if I'm missing something, where is the Type being `<20 x i64>` in this case? AFAICT, there isn't any load associated to the store when the store address is constant in x86, as it should just be needed to compute the memory address.

Comment (RKSimon) [reply llvm/lib/Target/X86/X86TargetTransformInfo.cpp:5162]: The cost you are disabling in the code above is for the constant load - not the memory address computation.

Comment (antoniofrighetto) [reply llvm/lib/Target/X86/X86TargetTransformInfo.cpp:5162]: That's correct. There is no cost for any constant load. There is only a memory address computation, which is _not_ a load.

Comment (RKSimon) [reply llvm/lib/Target/X86/X86TargetTransformInfo.cpp:5162]: That's only true for scalars - for vector constant it must be loaded from the constant pool unless rematerialized (which for x86 means 0 or -1 splats only). That is the issue in #111126 - the load from `LCPI0_0` is being overestimated (its just a <2 x i64> load that is reused in all 10 stores but we're treating it as a <20 x i64> load):
```
f():
        movaps  xmm0, xmmword ptr [rip + .LCPI0_0]
        movaps  xmmword ptr [rip + arr], xmm0
        movaps  xmmword ptr [rip + arr+16], xmm0
        movaps  xmmword ptr [rip + arr+32], xmm0
        movaps  xmmword ptr [rip + arr+48], xmm0
        movaps  xmmword ptr [rip + arr+64], xmm0
        movaps  xmmword ptr [rip + arr+80], xmm0
        movaps  xmmword ptr [rip + arr+96], xmm0
        movaps  xmmword ptr [rip + arr+112], xmm0
        movaps  xmmword ptr [rip + arr+128], xmm0
        movaps  xmmword ptr [rip + arr+144], xmm0
        ret
```
I'm investigating an alternative patch but have run out of time tonight :/

Comment (antoniofrighetto) [reply llvm/lib/Target/X86/X86TargetTransformInfo.cpp:5162]: Sorry, isn't this just bad codegen? In principle, we should be able to simply store the constant in the floating-point register, why would we need to load it from memory? GCC seems to catch this: https://godbolt.org/z/6baWMK4cW.

Comment (antoniofrighetto) [reply llvm/lib/Target/X86/X86TargetTransformInfo.cpp:5162]: Uhmm, we still need a load when dealing with arbitrary 64-bit constants.

Comment (antoniofrighetto) [reply llvm/lib/Target/X86/X86TargetTransformInfo.cpp:5162]: If any, the check would need to take this into account (scalar <= INT_MAX), but feel free to assess if there's a more favourable approach here.

**Comment by antoniofrighetto** [comment]
Closing this, as favoured approach in https://github.com/llvm/llvm-project/pull/118642.

**Pull request closed without merging**